bpf/lib/maglev.h

```diff
@@ -0,0 +1,153 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/* Copyright (C) 2016-2020 Authors of Cilium */
+
+#ifndef __MAGLEV_H_
+#define __MAGLEV_H_
```

Comment:
The PR description implies the ability to apply this maglev logic also to east-west traffic, not just north-south. @borkmann was there any particular motivation for why the current implementation doesn't address this case? Or is it just not the focus of the initially scoped work?

Reply:
We didn't add it to the E-W case which uses the socket LB given for TCP and connected UDP, we directly pick a backend once during connect() call which then the kernel uses for the rest of the socket lifetime. The unconnected UDP case could change backends if the order changes, but for that case there is already the session affinity to have the backends be sticky (https://docs.cilium.io/en/v1.8/gettingstarted/kubeproxy-free/#session-affinity). Either case, the socket LB doesn't go through any intermediate L4LB but always directly to the backend, hence we didn't add it there. Is there a rationale for the case where session affinity should not suffice for unconnected UDP?

Reply:
I checked the latest code, when `backend_id` is 0, just pick up a backend randomly (not maglev way). I didn't see the directly picking routine. Could someone can tell me what important routine I missed.
```
	if (backend_id == 0) {
		backend_from_affinity = false;

		key.backend_slot = (sock_select_slot(ctx_full) % svc->count) + 1;
		backend_slot = **__lb4_lookup_backend_slot**(&key);
		if (!backend_slot) {
			update_metrics(0, METRIC_EGRESS, REASON_LB_NO_BACKEND_SLOT);
			return -ENOENT;
		}

		backend_id = backend_slot->backend_id;
		backend = __lb4_lookup_backend(backend_id);
	}
```

Reply:
@AirbornWdd We don't use Maglev in that case. That's the case @borkmann was referring to as `E-W`:

> We didn't add it to the E-W case which uses the socket LB given for TCP and connected UDP, we directly pick a backend once during connect() call which then the kernel uses for the rest of the socket lifetime.